### PR TITLE
Bug #15229: Instances could be added to abstract types

### PR DESCRIPTION
**Pull request by fppt**

### CODE REVIEW

**Comment by fppt** [comment]
!rtg

File: grakn-graph/src/main/java/ai/grakn/graph/internal/ValidateGlobalRules.java [hunk 1]
```diff
@@ -189,7 +189,7 @@ private ValidateGlobalRules() {
                 conceptType.<TypeImpl>getIncomingNeighbours(Schema.EdgeLabel.SHARD).anyMatch(thing ->
                 thing.getIncomingNeighbours(Schema.EdgeLabel.ISA).findAny().isPresent())){
 
-            return Optional.of(VALIDATION_IS_ABSTRACT.getMessage(conceptType.getLabel()));
+            return Optional.of(IS_ABSTRACT.getMessage(conceptType.getLabel()));
         }
         return Optional.empty();
     }
```
Comment (aelred) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/ValidateGlobalRules.java]: Can we get rid of this validation rule?

File: grakn-graph/src/test/java/ai/grakn/generator/AbstractTypeGenerator.java [hunk 2]
```diff
@@ -92,11 +98,20 @@ private final boolean includeNonMeta() {
         return !meta.orElse(false);
     }
 
+    private final boolean includeAbstract(){
+        return isAbstract.orElse(false);
+    }
```
Comment (aelred) [inline grakn-graph/src/test/java/ai/grakn/generator/AbstractTypeGenerator.java]: This should be `.orElse(true)`. Essentially, if you don't specifically ask to include/exclude abstract types, then they should be included.

I'm wondering if there's a way to remove the `Optional` from `isAbstract` and `meta`, but I'm not sure...

File: grakn-graph/src/test/java/ai/grakn/graph/internal/ValidatorTest.java [hunk 3]
```diff
@@ -129,20 +130,12 @@ public void whenCommittingNonAbstractRelationTypeNotLinkedToAnyRoleType_Throw(){
...
-
-        x4.superType(x3);
-
-        expectedException.expect(GraknValidationException.class);
-        expectedException.expectMessage(containsString(ErrorMessage.VALIDATION_IS_ABSTRACT.getMessage(x1.getLabel())));
+        expectedException.expect(GraphRuntimeException.class);
+        expectedException.expectMessage(containsString(ErrorMessage.IS_ABSTRACT.getMessage(x1.getLabel())));
 
-        graknGraph.commit();
+        x1.setAbstract(true);
```
Comment (aelred) [inline grakn-graph/src/test/java/ai/grakn/graph/internal/ValidatorTest.java]: I'd delete this test, it seems the same as the above. The test name is also now incorrect.

File: grakn-test/src/test/java/ai/grakn/test/graql/query/InsertQueryTest.java [hunk 4]
```diff
@@ -417,7 +416,6 @@ public void testInsertReferenceByName() {
         //noinspection OptionalGetWithoutIsPresent
         EntityType newType = typeQuery.get("n").findFirst().get().asEntityType();
 
-        assertTrue(newType.asEntityType().isAbstract());
```
Comment (aelred) [inline grakn-test/src/test/java/ai/grakn/test/graql/query/InsertQueryTest.java]: Was this failing for some reason?

File: grakn-graph/src/main/java/ai/grakn/graph/internal/ValidateGlobalRules.java (hunk 1 above)
Comment (fppt) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/ValidateGlobalRules.java]: yes we can.

File: grakn-test/src/test/java/ai/grakn/test/graql/query/InsertQueryTest.java (hunk 4 above)
Comment (fppt) [inline grakn-test/src/test/java/ai/grakn/test/graql/query/InsertQueryTest.java]: Yes the test was attemping to add instances to an abstract type. So I stopped the type from being abstract which meant this line had to die.

File: grakn-graph/src/test/java/ai/grakn/graph/internal/ValidatorTest.java (hunk 3 above)
Comment (fppt) [inline grakn-graph/src/test/java/ai/grakn/graph/internal/ValidatorTest.java]: Which test above?

File: grakn-test/src/test/java/ai/grakn/test/graql/query/InsertQueryTest.java (hunk 4 above)
Comment (aelred) [inline grakn-test/src/test/java/ai/grakn/test/graql/query/InsertQueryTest.java]: Got it! Missed the `isa`.

**Pull request merged**